### PR TITLE
Refactored motionsensing

### PR DESCRIPTION
**Pull request by clare**
Yay, finally landed new motion sensing which compares with frames of several seconds ago instead of the last frame.  

We still need to do some real world testing with changing the min setting from 3000 -> 2900 but that is just in the config anyway.

### CODE REVIEW

File: thermal-recorder-TEMPLATE.yaml [hunk 1]
```diff
@@ -27,25 +30,41 @@ min-disk-space: 200
...
 
-# GPIO pins for the LEDs
-leds:
-    # Pin of the blue LED, indicating that the device is recording.
-    recording: "GPIO20"
+    # When working out which pixels have changed it tries to compare
+    # the current frame with one recorded this many frames ago. If there
+    # aren't this many frames (eg since a recalibration) then it will use
+    # the oldest frame it has after the last 'bad' frame.
+    FrameCompareGap: 45
```
Comment (mjs) [inline thermal-recorder-TEMPLATE.yaml]: incorrect name :)

File: thermal-recorder-TEMPLATE.yaml [hunk 2]
```diff
@@ -27,25 +30,41 @@ min-disk-space: 200
...
+
+    # When working out if motion has occured only look at a single frame diff.
+    # If set to false then it looks as two sequential frame diffs and only consideres
+    # pixels that have changed in both diffs
+    # Note: set to false because rats move too fast!
+    UseOneFrameOnly: true
```
Comment (mjs) [inline thermal-recorder-TEMPLATE.yaml]: same here

File: cmd/thermal-recorder/cptv_test.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
+// Copyright 2018 The Cacophony Project. All rights reserved.
+// Use of this source code is governed by the Apache License Version 2.0;
+// see the LICENSE file for further details.
```
Comment (mjs) [inline cmd/thermal-recorder/cptv_test.go]: Oh boy. This is wrong, but it's wrong throughout the thermal-recorder code. It's supposed to be GPLv3.

Can you just leave out the headers for now. I'll get this sorted in a later PR.

File: cmd/thermal-recorder/cptv_test.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func CurrentConfig() *Config {
+	dir := GetBaseDir()
+	config_file := strings.Replace(dir, "cmd/thermal-recorder", "thermal-recorder-TEMPLATE.yaml", 1)
+	uploader_config_file := dir + "/motiontest/thermal-uploader-test.yaml"
+	config, _ := ParseConfigFiles(config_file, uploader_config_file)
```
Comment (mjs) [inline cmd/thermal-recorder/cptv_test.go]: You shouldn't ignore any error that comes back here. Given that it's test code is perfectly OK to panic here.

File: cmd/thermal-recorder/cptv_test.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+func TestCptvAnimalRecordings(t *testing.T) {
+	config := CurrentConfig()
+
+	results := NewCPTVPlaybackTester(config).TestAllCPTVFiles(GetBaseDir() + "/motiontest/animals")
+
+	expectedResults := []string{
+		"cat.cptv             Detected: (25:41)          Recorded: (26:50)          Motion frames: 18/133",
+		"hedgehog.cptv        Detected: (3:32)(45:       Recorded: (4:41)(46:       Motion frames: 55/100",
```
Comment (mjs) [inline cmd/thermal-recorder/cptv_test.go]: What's with the truncated looking Detected and Recorded columns?

File: cmd/thermal-recorder/cptv_test.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+		"possum02.cptv        Detected: (1:              Recorded: (2:              Motion frames: 92/94",
+		"rat.cptv             Detected: (1:6)            Recorded: (2:15)           Motion frames: 7/99",
+		"rat02.cptv           Detected: (1:14)(61:92)    Recorded: (2:23)(62:101)   Motion frames: 39/107",
+		"recalc.cptv          Detected: (1:497)          Recorded: (2:506)          Motion frames: 490/540",
```
Comment (mjs) [inline cmd/thermal-recorder/cptv_test.go]: It may have been better to return the results as a slice of structs instead of a list of formatted strings (fragile).

Also, won't these tests break random ways in the face of almost any change - positive or negative - to the motion detection algorithm? Is that useful?

File: cmd/thermal-recorder/cptv_test.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+	f, err := os.Create(GetBaseDir() + "/motiontest/results/" + testname)
+	if err != nil {
+		return
+	}
+	defer f.Close()
+	writer := bufio.NewWriter(f)
+
+	config := CurrentConfig()
+	config.Motion.TempThresh = 2900
+	ExperiementAndWriteResultsToFile(testname+"2900", config, searchDir, writer)
```
Comment (mjs) [inline cmd/thermal-recorder/cptv_test.go]: Experiment

File: cmd/thermal-recorder/cptv_test.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+	fmt.Fprintln(writer)
+	fmt.Fprintln(writer)
+
+	writer.Flush()
+}
+
+func GetBaseDir() string {
+	_, file, _, _ := runtime.Caller(0)
+
+	dir, _ := filepath.Abs(filepath.Dir(file))
```
Comment (mjs) [inline cmd/thermal-recorder/cptv_test.go]: Again, don't ignore the error, panic if it happens. Panics are rarely a good idea for production code but are fine in test code.

File: cmd/thermal-recorder/cptv_test.go [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+	fmt.Fprintf(writer, "Results for %s", name)
+	fmt.Fprintln(writer)
+
+	results := NewCPTVPlaybackTester(config).TestAllCPTVFiles(dir)
+
+	fmt.Fprintf(writer, "%-10s:  %ds - %ds", "Recording limits", config.MinSecs, config.MaxSecs)
+	fmt.Fprintln(writer, "")
+	fmt.Fprintln(writer, config.Motion)
+
+	for ii := range results {
```
Comment (mjs) [inline cmd/thermal-recorder/cptv_test.go]: `for _, result := range results` is arguably nicer. It's more direct and avoids the short variable name.

File: cmd/thermal-recorder/cptv_test.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+	frames := tester.LoadAllCptvFrames(GetBaseDir() + "/motiontest/animals/recalc.cptv")
+
+	listener := new(HardwareListener) // currently doesn't do anything
+	recorder := new(NoWriteRecorder)
+
+	processor := NewMotionProcessor(config, listener, recorder)
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		for jj := range frames {
```
Comment (mjs) [inline cmd/thermal-recorder/cptv_test.go]: same here

File: cmd/thermal-recorder/frameloop.go [hunk 11]
```diff
@@ -61,24 +63,32 @@ func (fl *FrameLoop) Current() *lepton3.Frame {
 	return fl.frames[fl.currentIndex]
 }
 
-// Previous returns a copy of the previous frame.
-func (fl *FrameLoop) Previous() *lepton3.Frame {
+// CopyRecent returns a copy of the previous frame.
+func (fl *FrameLoop) CopyRecent(f *lepton3.Frame) *lepton3.Frame {
```
Comment (mjs) [inline cmd/thermal-recorder/frameloop.go]: Yep, that's a better name

File: cmd/thermal-recorder/frameloop.go [hunk 12]
```diff
@@ -33,9 +33,8 @@ type FrameLoop struct {
 	currentIndex  int
 	frames        []*lepton3.Frame
 	orderedFrames []*lepton3.Frame
-	zeroFrame     lepton3.Frame
 	bufferFull    bool
-	mu            sync.Mutex
+	oldest        int
```
Comment (mjs) [inline cmd/thermal-recorder/frameloop.go]: The mutex is gone but I don't see anything that provides goroutine safety now... what's up with that?

File: cmd/thermal-recorder/main.go [hunk 13]
```diff
@@ -39,8 +31,18 @@ type Args struct {
...
 	UploaderConfigFile string `arg:"-u,--uploader-config" help:"path to uploader config file"`
 	Timestamps         bool   `arg:"-t,--timestamps" help:"include timestamps in log output"`
+	TestCptvFile       string `arg:"-f, --testfile" help:"Run a CPTV file through to see what the results are"`
+	Verbose            bool   `arg:"-v, --verbose" help:"Make logging more verbose"`
 }
 
+type HardwareListener struct{}
+
+func (hl *HardwareListener) MotionDetected() {
+	//turret.Update(motion)
```
Comment (mjs) [inline cmd/thermal-recorder/main.go]: remove?

File: cmd/thermal-recorder/main.go [hunk 14]
```diff
@@ -169,81 +147,14 @@ func handleConn(conn net.Conn, conf *Config, turret *TurretController, recording
...
-		if writer != nil {
-			err := writer.WriteFrame(frame)
-			if err != nil {
-				return err
-			}
-			framesWritten++
-		}
-
-		frame = frameLoop.Move()
+		processor.Process(rawFrame)
```
Comment (mjs) [inline cmd/thermal-recorder/main.go:143]: great to have this pulled out, thanks

File: thermal-recorder-TEMPLATE.yaml [hunk 15]
```diff
@@ -27,25 +30,41 @@ min-disk-space: 200
...
-    # Pin of the blue LED, indicating that the device is recording.
-    recording: "GPIO20"
+    # When working out which pixels have changed it tries to compare
+    # the current frame with one recorded this many frames ago. If there
+    # aren't this many frames (eg since a recalibration) then it will use
+    # the oldest frame it has after the last 'bad' frame.
+    FrameCompareGap: 45
+
+    # When working out if motion has occured only look at a single frame diff.
+    # If set to false then it looks as two sequential frame diffs and only consideres
```
Comment (mjs) [inline thermal-recorder-TEMPLATE.yaml]: considers

File: thermal-recorder-TEMPLATE.yaml [hunk 16]
```diff
@@ -27,25 +30,41 @@ min-disk-space: 200
...
+    # pixels that have changed in both diffs
+    # Note: set to false because rats move too fast!
```
Comment (mjs) [inline thermal-recorder-TEMPLATE.yaml]: I don't understand this note. It seems to be saying that it is preferable to have this setting set to false, but the default is true.

File: thermal-recorder-TEMPLATE.yaml (hunk 2 above)
Comment (mjs) [reply thermal-recorder-TEMPLATE.yaml]: Would a better name for this "one-diff-only"?

File: cmd/thermal-recorder/cptv_test.go [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+	return dir
+}
+
+func BenchmarkMotionDetection(b *testing.B) {
+	config := CurrentConfig()
+
+	tester := NewCPTVPlaybackTester(config)
+	frames := tester.LoadAllCptvFrames(GetBaseDir() + "/motiontest/animals/recalc.cptv")
+
+	listener := new(HardwareListener) // currently doesn't do anything
```
Comment (mjs) [inline cmd/thermal-recorder/cptv_test.go]: Although the HardwareListener doesn't do anything, there's no need for the comment. Also, the comment becomes incorrect as soon as the HardwareListener changes.

File: cmd/thermal-recorder/cptv_test.go [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+			errors++
+		} else if expected != actual.motionDetectedFrames {
+			log.Printf("Expected results for %-16s: %s", key, expected)
+			log.Printf("Actual results for   %-16s: %s", key, actual.motionDetectedFrames)
+			errors++
+		}
+	}
+
+	for key, result := range actual {
+		if _, ok := expectedResults[key]; ok == false {
```
Comment (mjs) [inline cmd/thermal-recorder/cptv_test.go:76]: `!ok` instead of `ok == false`

File: cmd/thermal-recorder/cptv_test.go [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+	actualResults := NewCPTVPlaybackTester(config).TestAllCPTVFiles(GetBaseDir() + "/motiontest/animals")
+
+	expectedResults := map[string]string{
+		"cat.cptv":      "(25:41)",
+		"hedgehog.cptv": "(3:32)(45:end)",
+		"possum02.cptv": "(1:end)",
+		"rat.cptv":      "(1:6)",
+		"rat02.cptv":    "(1:14)(61:92)",
+		"recalc.cptv":   "(1:497)",
+	}
```
Comment (mjs) [inline cmd/thermal-recorder/cptv_test.go:56]: Thanks for changing this. Much better.

File: cmd/thermal-recorder/cptv_test.go [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+
+	results := NewCPTVPlaybackTester(config).TestAllCPTVFiles(dir)
+
+	fmt.Fprintf(writer, "%-10s:  %ds - %ds", "Recording limits", config.MinSecs, config.MaxSecs)
+	fmt.Fprintln(writer)
+	fmt.Fprintf(writer, "Motion: %+v", config.Motion)
+	fmt.Fprintln(writer)
+
+	for key, result := range results {
+		fmt.Fprintf(writer, "%-20s Detected: %-16s Recorded: %-16s Motion frames: %d/%d", key, result.motionDetectedFrames, result.recordedFrames, result.motionDetectedCount, result.frameCount)
```
Comment (mjs) [inline cmd/thermal-recorder/cptv_test.go]: This would read better if the line was split into multiple lines

File: cmd/thermal-recorder/main.go [hunk 21]
```diff
@@ -39,7 +27,17 @@ type Args struct {
...
 	UploaderConfigFile string `arg:"-u,--uploader-config" help:"path to uploader config file"`
 	Timestamps         bool   `arg:"-t,--timestamps" help:"include timestamps in log output"`
+	TestCptvFile       string `arg:"-f, --testfile" help:"Run a CPTV file through to see what the results are"`
+	Verbose            bool   `arg:"-v, --verbose" help:"Make logging more verbose"`
+}
+
+type HardwareListener struct{}
+
+func (hl *HardwareListener) MotionDetected() {
+	//turret.Update(motion)
```
Comment (mjs) [inline cmd/thermal-recorder/main.go]: remove commented out line?

**Review by mjs — APPROVED** [review]
Looking good. Just a couple of minor things to look at and this is good to merge.
